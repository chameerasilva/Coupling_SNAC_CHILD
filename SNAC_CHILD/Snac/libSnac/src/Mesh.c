/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
**
** Copyright (C), 2003, 
**	Steve Quenette, 110 Victoria Street, Melbourne, Victoria, 3053, Australia.
**	Californian Institute of Technology, 1200 East California Boulevard, Pasadena, California, 91125, USA.
**	University of Texas, 1 University Station, Austin, Texas, 78712, USA.
**
** Authors:
**	Stevan M. Quenette, Senior Software Engineer, VPAC. (steve@vpac.org)
**	Stevan M. Quenette, Visitor in Geophysics, Caltech.
**	Luc Lavier, Research Scientist, The University of Texas. (luc@utig.ug.utexas.edu)
**	Luc Lavier, Research Scientist, Caltech.
**
** This program is free software; you can redistribute it and/or modify it
** under the terms of the GNU General Public License as published by the
** Free Software Foundation; either version 2, or (at your option) any
** later version.
** 
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
** 
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
**
** $Id: Mesh.c 3274 2007-03-27 20:25:29Z EunseoChoi $
**
**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

#include <mpi.h>
#include <StGermain/StGermain.h>
#include <StGermain/FD/FD.h>

#include "types.h"
#include "units.h"
#include "Material.h"
#include "Context.h"
#include "Mesh.h"
#include <stdio.h>
#include <string.h>


void Snac_Mesh_Delete( void* mesh ) {
	Snac_Mesh*	self = (Snac_Mesh*)mesh;
	
	if( self->force ) {
		Memory_Free( self->force );
		self->force = 0;
	}
}

void Snac_Mesh_Print( void* mesh, Stream* stream ) {
	Snac_Mesh*	self = (Snac_Mesh*)mesh;
	Node_LocalIndex	node_lI;
	
	Journal_Printf( stream, "Snac_Mesh:\n" );
	
	Journal_Printf( stream, "\tnodeLocalCount: %u\n", self->nodeLocalCount );
}

void Snac_Mesh_Build( void* mesh, void* _context ) {
	Snac_Mesh*	self = (Snac_Mesh*)mesh;
	Snac_Context*	context = (Snac_Context*)_context;
	
	self->debug = context->debug;
	self->rank = context->rank;
	Journal_Printf( self->debug, "In: %s\n", __func__ );
	
	self->nodeLocalCount = context->mesh->nodeLocalCount;
}

void Snac_Mesh_IC( void* mesh ) {
	Snac_Mesh*	self = (Snac_Mesh*)mesh;
	
	Journal_Printf( self->debug, "In: %s\n", __func__ );
}
