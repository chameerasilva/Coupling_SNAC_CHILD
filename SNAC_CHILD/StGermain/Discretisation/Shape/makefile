
#Finds the Absolute path to the Project Root directory
SHELL := /bin/bash
PROJ_ROOT := $(shell until test -r ./Makefile.system ; do cd .. ; done ; echo `pwd`)
include ${PROJ_ROOT}/Makefile.system

# Subdirectories
subdirs := src tests

include ${PROJ_ROOT}/Makefile.vmake
