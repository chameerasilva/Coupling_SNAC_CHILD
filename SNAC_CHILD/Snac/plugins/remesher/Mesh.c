/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
**
** Copyright (C), 2003,
**	Steve Quenette, 110 Victoria Street, Melbourne, Victoria, 3053, Australia.
**	Californian Institute of Technology, 1200 East California Boulevard, Pasadena, California, 91125, USA.
**	University of Texas, 1 University Station, Austin, Texas, 78712, USA.
**
** Authors:
**	Stevan M. Quenette, Senior Software Engineer, VPAC. (steve@vpac.org)
**	Stevan M. Quenette, Visitor in Geophysics, Caltech.
**	Luc Lavier, Research Scientist, The University of Texas. (luc@utig.ug.utexas.edu)
**	Luc Lavier, Research Scientist, Caltech.
**
** This program is free software; you can redistribute it and/or modify it
** under the terms of the GNU General Public License as published by the
** Free Software Foundation; either version 2, or (at your option) any
** later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
**
** $Id: Mesh.c 3095 2005-07-13 09:50:46Z LukeHodkinson $
**
**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

#include <mpi.h>
#include <StGermain/StGermain.h>
#include <StGermain/FD/FD.h>

#include "Snac/Snac.h"
#include "types.h"
#include "Mesh.h"

#include <stdio.h>


/*
** Due to the rewrite this func is severely useless.  Need to fix it.
*/

void SnacRemesher_Mesh_Print( void* mesh, Stream* stream ) {
	SnacRemesher_Mesh*	self = (SnacRemesher_Mesh*)mesh;
/*	Node_LocalIndex		node_lI; */
/*	Node_LocalIndex		index; */

	Journal_Printf( stream, "SnacRemesher_Mesh:\n" );

#if 0
	Journal_Printf( stream, "\tnodeLocalCount: %u\n", self->nodeLocalCount );
	Journal_Printf( stream, "\tnodeDomainCount: %u\n", self->nodeDomainCount );
	Journal_Printf( stream, "\tnodeGlobalCount: %u\n", self->nodeGlobalCount );
	Journal_Printf( stream, "\telementLocalCount: %u\n", self->elementLocalCount );
	Journal_Printf( stream, "\telementDomainCount: %u\n", self->elementDomainCount );
	Journal_Printf( stream, "\telementGlobalCount: %u\n", self->elementGlobalCount );
	Journal_Printf( stream, "\tinitialNodeCoord[0-%u]: { ", self->nodeLocalCount );
	for( node_lI = 0; node_lI < self->nodeLocalCount; node_lI++ ) {
		Journal_Printf( stream,
			"{ %g, %g, %g } ",
			self->initialNodeCoord[node_lI][0],
			self->initialNodeCoord[node_lI][1],
			self->initialNodeCoord[node_lI][2] );
	}
	Journal_Printf( stream, "}\n" );
	Journal_Printf( stream, "\tnewNodeCoord[0-%u]: { ", self->nodeLocalCount );
	for( node_lI = 0; node_lI < self->nodeLocalCount; node_lI++ ) {
		Journal_Printf( stream,
			"{ %g, %g, %g } ",
			self->newNodeCoord[node_lI][0],
			self->newNodeCoord[node_lI][1],
			self->newNodeCoord[node_lI][2] );
	}
	Journal_Printf( stream, "}\n" );
	Journal_Printf( stream, "\tnewNode[0-%u]: { ... TODO: Print nodes nicely... }\n", self->nodeLocalCount );

	Journal_Printf( stream,
		"\tmeshType: %s\n",
		self->meshType == SnacRemesher_Spherical ? "Spherical" :
			self->meshType == SnacRemesher_Cartesian ? "Cartesian" : "Invalid value!" );

	Journal_Printf( stream, "\tneighbourRankCount: %u\n", self->neighbourRankCount );
	Journal_Printf( stream, "\tneighbourRank[0-%u]: {", self->neighbourRankCount );
	for( index = 0; index < self->neighbourRankCount; index++ ) {
		Journal_Printf( stream, "%i ", self->neighbourRank[index] );
	}
	Journal_Printf( stream, "}\n" );

	Journal_Printf( stream, "\ttopInternalSet: " ); Print( self->topInternalSet, stream );
	Journal_Printf( stream, "\tbottomInternalSet: " ); Print( self->bottomInternalSet, stream );
	Journal_Printf( stream, "\twallSet: " ); Print( self->wallSet, stream );
	Journal_Printf( stream, "\tinternalSet: " ); Print( self->internalSet, stream );

	Journal_Printf( stream, "\ttopInternalCount: %u\n", self->topInternalCount );
	Journal_Printf( stream, "\tbottomInternalCount: %u\n", self->bottomInternalCount );
	Journal_Printf( stream, "\twalllCount: %u\n", self->wallCount );
	Journal_Printf( stream, "\tinternalCount: %u\n", self->internalCount );

	Journal_Printf( stream, "\ttopInternalArray[0-%u]: { ", self->topInternalCount );
	for( index = 0; index < self->topInternalCount; index++ ) {
		Journal_Printf( stream, "%u ", self->topInternalArray[index] );
	}
	Journal_Printf( stream, "}\n" );
	Journal_Printf( stream, "\tbottomInternalArray[0-%u]: { ", self->bottomInternalCount );
	for( index = 0; index < self->bottomInternalCount; index++ ) {
		Journal_Printf( stream, "%u ", self->bottomInternalArray[index] );
	}
	Journal_Printf( stream, "}\n" );
	Journal_Printf( stream, "\twallSet[0-%u]: { ", self->wallCount );
	for( index = 0; index < self->wallCount; index++ ) {
		Journal_Printf( stream, "%u ", self->wallArray[index] );
	}
	Journal_Printf( stream, "}\n" );
	Journal_Printf( stream, "\tinternalArray[0-%u]: { ", self->internalCount );
	for( index = 0; index < self->internalCount; index++ ) {
		Journal_Printf( stream, "%u ", self->internalArray[index] );
	}
	Journal_Printf( stream, "}\n" );

	/* Shadow */
	Journal_Printf( stream, "\tshadowNodeLocalCount[0-%u]: { ", self->neighbourRankCount );
	for( index = 0; index < self->neighbourRankCount; index++ ) {
		Journal_Printf( stream, "%u ", self->shadowNodeLocalCount[index] );
	}
	Journal_Printf( stream, "}\n" );
	Journal_Printf( stream, "\tshadowNodeLocalArray[0-%u]: {\n", self->neighbourRankCount );
	for( index = 0; index < self->neighbourRankCount; index++ ) {
		Index					index2;

		Journal_Printf( stream, "\t\tshadowNodeLocalArray[%u][0-%u]: { ", index, self->shadowNodeLocalCount[index] );
		for( index2 = 0; index2 < self->shadowNodeLocalCount[index]; index2++ ) {
			Journal_Printf( stream, "%u ", self->shadowNodeLocalArray[index][index2] );
		}
		Journal_Printf( stream, "}\n" );
	}
	Journal_Printf( stream, "}\n" );
	Journal_Printf( stream, "\tshadowNodeRemoteCount[0-%u]: { ", self->neighbourRankCount );
	for( index = 0; index < self->neighbourRankCount; index++ ) {
		Journal_Printf( stream, "%u ", self->shadowNodeRemoteCount[index] );
	}
	Journal_Printf( stream, "}\n" );
	Journal_Printf( stream, "\tshadowNodeRemoteArray[0-%u]: {\n", self->neighbourRankCount );
	for( index = 0; index < self->neighbourRankCount; index++ ) {
		Index					index2;

		Journal_Printf( stream, "\t\tshadowNodeRemoteArray[%u][0-%u]: { ", index, self->shadowNodeRemoteCount[index] );
		for( index2 = 0; index2 < self->shadowNodeRemoteCount[index]; index2++ ) {
			Journal_Printf( stream, "%u ", self->shadowNodeRemoteArray[index][index2] );
		}
		Journal_Printf( stream, "}\n" );
	}
	Journal_Printf( stream, "}\n" );
	Journal_Printf( stream, "\tshadowNodeRemoteCoord[0-%u]: {\n", self->neighbourRankCount );
	for( index = 0; index < self->neighbourRankCount; index++ ) {
		Index					index2;

		Journal_Printf( stream, "\t\tshadowNodeRemoteCoord[%u][0-%u]: { ", index, self->shadowNodeRemoteCount[index] );
		for( index2 = 0; index2 < self->shadowNodeRemoteCount[index]; index2++ ) {
			Journal_Printf( stream,
				"{ %g, %g, %g }, ",
				self->shadowNodeRemoteCoord[index][index2][0],
				self->shadowNodeRemoteCoord[index][index2][1],
				self->shadowNodeRemoteCoord[index][index2][2] );
		}
		Journal_Printf( stream, "}\n" );
	}
	Journal_Printf( stream, "}\n" );
	Journal_Printf( stream, "\tshadowNodeRemote[0-%u]: {\n", self->neighbourRankCount );
	for( index = 0; index < self->neighbourRankCount; index++ ) {
		Index					index2;

		Journal_Printf( stream, "\t\tshadowNodeRemote[%u][0-%u]: { ", index, self->shadowNodeRemoteCount[index] );
		for( index2 = 0; index2 < self->shadowNodeRemoteCount[index]; index2++ ) {
			Snac_Node_Print( &self->shadowNodeRemote[index][index2] );
		}
		Journal_Printf( stream, "}\n" );
	}
	Journal_Printf( stream, "}\n" );

	/* Shadow Elements*/
	Journal_Printf( stream, "\tshadowElementLocalCount[0-%u]: { ", self->neighbourRankCount );
	for( index = 0; index < self->neighbourRankCount; index++ ) {
		Journal_Printf( stream, "%u ", self->shadowElementLocalCount[index] );
	}
	Journal_Printf( stream, "}\n" );
	Journal_Printf( stream, "\tshadowElementLocalArray[0-%u]: {\n", self->neighbourRankCount );
	for( index = 0; index < self->neighbourRankCount; index++ ) {
		Index					index2;

		Journal_Printf( stream, "\t\tshadowElementLocalArray[%u][0-%u]: { ", index, self->shadowElementLocalCount[index] );
		for( index2 = 0; index2 < self->shadowElementLocalCount[index]; index2++ ) {
			Journal_Printf( stream, "%u ", self->shadowElementLocalArray[index][index2] );
		}
		Journal_Printf( stream, "}\n" );
	}
	Journal_Printf( stream, "}\n" );
	Journal_Printf( stream, "\tshadowElementRemoteCount[0-%u]: { ", self->neighbourRankCount );
	for( index = 0; index < self->neighbourRankCount; index++ ) {
		Journal_Printf( stream, "%u ", self->shadowElementRemoteCount[index] );
	}
	Journal_Printf( stream, "}\n" );
	Journal_Printf( stream, "\tshadowElementRemoteArray[0-%u]: {\n", self->neighbourRankCount );
	for( index = 0; index < self->neighbourRankCount; index++ ) {
		Index					index2;

		Journal_Printf( stream, "\t\tshadowElementRemoteArray[%u][0-%u]: { ", index, self->shadowElementRemoteCount[index] );
		for( index2 = 0; index2 < self->shadowElementRemoteCount[index]; index2++ ) {
			Journal_Printf( stream, "%u ", self->shadowElementRemoteArray[index][index2] );
		}
		Journal_Printf( stream, "}\n" );
	}
	Journal_Printf( stream, "}\n" );
	for( index = 0; index < self->neighbourRankCount; index++ ) {
		Index					index2;

		Journal_Printf( stream, "\t\tshadowElementRemote[%u][0-%u]: { ", index, self->shadowElementRemoteCount[index] );
		for( index2 = 0; index2 < self->shadowElementRemoteCount[index]; index2++ ) {
			Snac_Element_Print( &self->shadowElementRemote[index][index2], stream );
		}
		Journal_Printf( stream, "}\n" );
	}
	Journal_Printf( stream, "}\n" );
#endif
}
