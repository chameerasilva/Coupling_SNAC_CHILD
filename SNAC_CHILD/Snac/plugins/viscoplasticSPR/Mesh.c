/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
**
** Copyright (C), 2003,
**	Steve Quenette, 110 Victoria Street, Melbourne, Victoria, 3053, Australia.
**	Californian Institute of Technology, 1200 East California Boulevard, Pasadena, California, 91125, USA.
**	University of Texas, 1 University Station, Austin, Texas, 78712, USA.
**
** Authors:
**	Stevan M. Quenette, Senior Software Engineer, VPAC. (steve@vpac.org)
**	Stevan M. Quenette, Visitor in Geophysics, Caltech.
**	Luc Lavier, Research Scientist, The University of Texas. (luc@utig.ug.utexas.edu)
**	Luc Lavier, Research Scientist, Caltech.
**
** This program is free software; you can redistribute it and/or modify it
** under the terms of the GNU General Public License as published by the
** Free Software Foundation; either version 2, or (at your option) any
** later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
**
** $Id: Mesh.c 3095 2005-07-13 09:50:46Z LukeHodkinson $
**
**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

#include <mpi.h>
#include <StGermain/StGermain.h>
#include <StGermain/FD/FD.h>

#include "Snac/Snac.h"
#include "types.h"
#include "Mesh.h"

#include <stdio.h>


/*
** Due to the rewrite this func is severely useless.  Need to fix it.
*/

void SnacViscoPlastic_Mesh_Print( void* mesh, Stream* stream ) {
	SnacViscoPlastic_Mesh*	self = (SnacViscoPlastic_Mesh*)mesh;

	Journal_Printf( stream, "SnacViscoPlastic_Mesh:\n" );

}
