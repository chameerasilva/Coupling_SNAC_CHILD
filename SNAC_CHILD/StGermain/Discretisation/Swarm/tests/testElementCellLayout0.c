/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
**
** Copyright (C), 2003, Victorian Partnership for Advanced Computing (VPAC) Ltd, 110 Victoria Street, Melbourne, 3053, Australia.
**
** Authors:
**	Stevan M. Quenette, Senior Software Engineer, VPAC. (steve@vpac.org)
**	Patrick D. Sunter, Software Engineer, VPAC. (pds@vpac.org)
**	Luke J. Hodkinson, Computational Engineer, VPAC. (lhodkins@vpac.org)
**	Siew-Ching Tan, Software Engineer, VPAC. (siew@vpac.org)
**	Alan H. Lo, Computational Engineer, VPAC. (alan@vpac.org)
**	Raquibul Hassan, Computational Engineer, VPAC. (raq@vpac.org)
**
**  This library is free software; you can redistribute it and/or
**  modify it under the terms of the GNU Lesser General Public
**  License as published by the Free Software Foundation; either
**  version 2.1 of the License, or (at your option) any later version.
**
**  This library is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
**  Lesser General Public License for more details.
**
**  You should have received a copy of the GNU Lesser General Public
**  License along with this library; if not, write to the Free Software
**  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
**
** Role:
**	Test that the ElementCellLayout has the same layout and geometry as the mesh's element layout.
**
** Assumptions:
**	None as yet.
**
** Comments:
**	None as yet.
**
** $Id: testElementCellLayout0.c 3555 2006-05-10 07:05:46Z PatrickSunter $
**
**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

#include <mpi.h>
#include "Base/Base.h"

#include "Discretisation/Geometry/Geometry.h"
#include "Discretisation/Shape/Shape.h"
#include "Discretisation/Mesh/Mesh.h"
#include "Discretisation/Utils/Utils.h"
#include "Discretisation/Swarm/Swarm.h"

#include <stdio.h>
#include <stdlib.h>

#define ConvertNode(node) \
	(node == 2 ? 3 : node == 3 ? 2 : node == 6 ? 7 : node == 7 ? 6 : node)

struct _Node {
	double temp;
};

struct _Element {
	double temp;
};

int main( int argc, char* argv[] ) {
	MPI_Comm			CommWorld;
	int				rank;
	int				numProcessors;
	int				procToWatch;
	Dictionary*			dictionary;
	Topology*			nTopology;
	ElementLayout*			eLayout;
	NodeLayout*			nLayout;
	MeshDecomp*			decomp;
	MeshLayout*			layout;
	ExtensionManager_Register*		extensionMgr_Register;
	Mesh*				mesh;
	ElementCellLayout*		elementCellLayout;
	
	/* Initialise MPI, get world info */
	MPI_Init( &argc, &argv );
	MPI_Comm_dup( MPI_COMM_WORLD, &CommWorld );
	MPI_Comm_size( CommWorld, &numProcessors );
	MPI_Comm_rank( CommWorld, &rank );
	
	Base_Init( &argc, &argv );
	
	DiscretisationGeometry_Init( &argc, &argv );
	DiscretisationShape_Init( &argc, &argv );
	DiscretisationMesh_Init( &argc, &argv );
	DiscretisationUtils_Init( &argc, &argv );
	DiscretisationSwarm_Init( &argc, &argv );
	MPI_Barrier( CommWorld ); /* Ensures copyright info always come first in output */

	if( argc >= 2 ) {
		procToWatch = atoi( argv[1] );
	}
	else {
		procToWatch = 0;
	}
	if( rank == procToWatch ) printf( "Watching rank: %i\n", rank );
	
	/* Read input */
	dictionary = Dictionary_New();
	Dictionary_Add( dictionary, "rank", Dictionary_Entry_Value_FromUnsignedInt( rank ) );
	Dictionary_Add( dictionary, "numProcessors", Dictionary_Entry_Value_FromUnsignedInt( numProcessors ) );
	Dictionary_Add( dictionary, "meshSizeI", Dictionary_Entry_Value_FromUnsignedInt( 3 ) );
	Dictionary_Add( dictionary, "meshSizeJ", Dictionary_Entry_Value_FromUnsignedInt( 4 ) );
	Dictionary_Add( dictionary, "meshSizeK", Dictionary_Entry_Value_FromUnsignedInt( 3 ) );
	Dictionary_Add( dictionary, "allowUnbalancing", Dictionary_Entry_Value_FromBool( True ) );
	Dictionary_Add( dictionary, "buildElementNodeTbl", Dictionary_Entry_Value_FromBool( True ) );
	Dictionary_Add( dictionary, "minX", Dictionary_Entry_Value_FromDouble( 0.0f ) );
	Dictionary_Add( dictionary, "minY", Dictionary_Entry_Value_FromDouble( 0.0f ) );
	Dictionary_Add( dictionary, "minZ", Dictionary_Entry_Value_FromDouble( 0.0f ) );
	Dictionary_Add( dictionary, "maxX", Dictionary_Entry_Value_FromDouble( 300.0f ) );
	Dictionary_Add( dictionary, "maxY", Dictionary_Entry_Value_FromDouble( 12.0f ) );
	Dictionary_Add( dictionary, "maxZ", Dictionary_Entry_Value_FromDouble( 300.0f ) );
	
	/* Run the mesher */
	nTopology = (Topology*)IJK6Topology_New( "IJK6Topology", dictionary );
	eLayout = (ElementLayout*)ParallelPipedHexaEL_New( "PPHexaEL", 3, dictionary );
	nLayout = (NodeLayout*)CornerNL_New( "CornerNL", dictionary, eLayout, nTopology );
	decomp = (MeshDecomp*)HexaMD_New( "HexaMD", dictionary, MPI_COMM_WORLD, eLayout, nLayout );
	layout = MeshLayout_New( "MeshLayout", eLayout, nLayout, decomp );
	
	/* Init mesh */
	extensionMgr_Register = ExtensionManager_Register_New();
	mesh = Mesh_New( "Mesh", layout, sizeof(Node), sizeof(Element), extensionMgr_Register, dictionary );
	
	/* Configure the element-cell-layout */
	elementCellLayout = ElementCellLayout_New( "elementCellLayout", mesh );
	
	/* Build the mesh */
	Build( mesh, 0, False );
	Initialise( mesh, 0, False );
	
	if( rank == procToWatch ) {
		Cell_Index cell;
		Element_DomainIndex	element;
		GlobalParticle          testParticle;
		
		for( element = 0; element < mesh->elementLocalCount; element++ ) {
			Cell_PointIndex			point;
			Cell_PointIndex			count;
			Coord**				cellPoints;
			Bool				result;

			cell = CellLayout_MapElementIdToCellId( elementCellLayout, element );

			if ( cell != element ) { 
				printf( "Wrong result: CellLayout_MapElementIdToCellId returned %d, when %d "
					"expected.\n", cell, element );
				exit(0);
			}

			count = elementCellLayout->_pointCount( elementCellLayout, cell );
			printf( "cellPointTbl  [%2u][0-%u]:\n", cell, count );
			cellPoints = Memory_Alloc_Array( Coord*, count, "cellPoints" );
			elementCellLayout->_initialisePoints( elementCellLayout, cell, count, cellPoints );
			for( point = 0; point < count; point++ ) {
				printf( "\t{%.3g %.3g %.3g}\n", (*cellPoints[ConvertNode(point)])[0], (*cellPoints[ConvertNode(point)])[1], 
					(*cellPoints[ConvertNode(point)])[2] );
			}
			printf( "\n" );

			testParticle.coord[0] = ( (*cellPoints[0])[0] + (*cellPoints[1])[0] ) / 2;
			testParticle.coord[1] = ( (*cellPoints[0])[1] + (*cellPoints[3])[1] ) / 2;
			testParticle.coord[2] = ( (*cellPoints[0])[2] + (*cellPoints[4])[2] ) / 2;
			printf( "Testing if test particle at (%f,%f,%f) is in the cell: ",
				testParticle.coord[0], testParticle.coord[1], testParticle.coord[2] );
			result = CellLayout_IsInCell( elementCellLayout, cell, &testParticle );
			printf( "%d\n\n", result );

			testParticle.coord[0] = (*cellPoints[count-1])[0] + 1;
			testParticle.coord[1] = (*cellPoints[count-1])[1] + 1;
			testParticle.coord[2] = (*cellPoints[count-1])[2] + 1;
			printf( "Testing if test particle at (%f,%f,%f) is in the cell: ",
				testParticle.coord[0], testParticle.coord[1], testParticle.coord[2] );
			result = CellLayout_IsInCell( elementCellLayout, cell, &testParticle );
			printf( "%d\n\n", result );

			Memory_Free( cellPoints );
		}
	}
	
	/* Destroy stuff */
	Stg_Class_Delete( elementCellLayout );
	Stg_Class_Delete( mesh );
	Stg_Class_Delete( extensionMgr_Register );
	Stg_Class_Delete( layout );
	Stg_Class_Delete( decomp );
	Stg_Class_Delete( nLayout );
	Stg_Class_Delete( eLayout );
	Stg_Class_Delete( nTopology );
	Stg_Class_Delete( dictionary );
	
	DiscretisationSwarm_Finalise();
	DiscretisationUtils_Finalise();
	DiscretisationMesh_Finalise();
	DiscretisationShape_Finalise();
	DiscretisationGeometry_Finalise();
	
	Base_Finalise();
	
	/* Close off MPI */
	MPI_Finalize();
	
	return 0; /* success */
}
