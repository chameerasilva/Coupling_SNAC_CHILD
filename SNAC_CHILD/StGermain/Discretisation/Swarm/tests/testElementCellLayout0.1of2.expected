StGermain Framework. Copyright (C) 2003-2005 VPAC.
Watching rank: 1
cellPointTbl  [ 0][0-8]:
	{0 8 0}
	{150 8 0}
	{0 12 0}
	{150 12 0}
	{0 8 150}
	{150 8 150}
	{0 12 150}
	{150 12 150}

Testing if test particle at (75.000000,10.000000,75.000000) is in the cell: 1

Testing if test particle at (1.000000,13.000000,151.000000) is in the cell: 0

cellPointTbl  [ 1][0-8]:
	{150 8 0}
	{300 8 0}
	{150 12 0}
	{300 12 0}
	{150 8 150}
	{300 8 150}
	{150 12 150}
	{300 12 150}

Testing if test particle at (225.000000,10.000000,75.000000) is in the cell: 1

Testing if test particle at (151.000000,13.000000,151.000000) is in the cell: 0

cellPointTbl  [ 2][0-8]:
	{0 8 150}
	{150 8 150}
	{0 12 150}
	{150 12 150}
	{0 8 300}
	{150 8 300}
	{0 12 300}
	{150 12 300}

Testing if test particle at (75.000000,10.000000,225.000000) is in the cell: 1

Testing if test particle at (1.000000,13.000000,301.000000) is in the cell: 0

cellPointTbl  [ 3][0-8]:
	{150 8 150}
	{300 8 150}
	{150 12 150}
	{300 12 150}
	{150 8 300}
	{300 8 300}
	{150 12 300}
	{300 12 300}

Testing if test particle at (225.000000,10.000000,225.000000) is in the cell: 1

Testing if test particle at (151.000000,13.000000,301.000000) is in the cell: 0

